docs/design_proposals/design-2_ignored_variables.md

```diff
@@ -0,0 +1,93 @@
+
+The current design of the interface for creating a posterior distribution does not allow for variables to be ignored in this manner. For example, if we use the Gaussian mean-field posterior, the logic is as follows:
+
+```python
+def create_Gaussian_meanfield(model, observed, dtype=None):
+    dtype = get_default_dtype() if dtype is None else dtype
+    observed = variables_to_UUID(observed)
+    q = Posterior(model)
+    for v in model.variables.values():
+        if v.type == VariableType.RANDVAR and v not in observed:
+            mean = Variable(shape=v.shape)
+            variance = Variable(shape=v.shape,
+                                transformation=PositiveTransformation())
+            q[v].set_prior(Normal(mean=mean, variance=variance, dtype=dtype))
+    return q
+```
+
+For the model depicted above, we have the following pattern of observations:
+
+```python
+observed = [model.x, model.y0]
+```
+ 
+and we end up with an additional variable for the head not being used (e.g. `y1` when observing `y0`). This then causes issues in the subsequent inference.
+ 
+
+## Proposed Changes
+
+One way to solve this is to have a to pass in a list of variables that are to be ignored when creating the posteriors, and then another list of variables. 
+As an example, for the Gaussian mean-field posterior example we could do the following: 
+
+```python
+def create_Gaussian_meanfield(model, observed, ignored=None, dtype=None):
+    dtype = get_default_dtype() if dtype is None else dtype
+    observed = variables_to_UUID(observed)
+    ignored = variables_to_UUID(ignored) if ignored is not None else []
+    q = Posterior(model)
+    for v in model.variables.values():
+        if v.type == VariableType.RANDVAR and v not in observed and v not in ignored:
+            mean = Variable(shape=v.shape)
+            variance = Variable(shape=v.shape,
+                                transformation=PositiveTransformation())
+            q[v].set_prior(Normal(mean=mean, variance=variance, dtype=dtype))
+    return q
+```
+
+We would then specify:
+
+```python
+observed = [model.x, model.y0]
+ignored = [model.y1]
+```
+
+Similarly, for the inference, would augment the keyword arguments to include these variables. i.e.:
+
+```python
+kwargs = dict(x=x, y0=y, ignored=[model.y1, model.r1])
+inference.run(max_iter=max_iter, learning_rate=learning_rate, verbose=False, callback=print_status, **kwargs)
+```
+
+Note here that we have additionally specified that the parent of `y1`, `r1`, should also be ignored. In terms of the Machine Learning algorithm, this would then be estimating the posterior:
+
+```
+p(Θ, r0, r1 | x, y0, y1) = p(Θ, r0 | x, y0) ∝ p(x, y0 , r0| Θ) p(Θ) 
+```
+
+where `Θ` denotes all of the weights and biases collected together.
+
+## Rejected Alternatives
+
```

Comment:
Is there a reason you can't pass a "targets" list, instead of "ignored"? I think we could then use those targets to only compute log_pdf / draw samples for those variables (and their associated dependencies). 

Also it's not really an issue that we define mean-field for all variables in the posterior definition, as long as we ignore unnecessary parts during the computation. This I feel less strongly about as it doesn't change the main API, only the create_gaussian_meanfield helper function.

So in the above, it would look something like:

```python
kwargs = dict(x=x, y0=y, targets=[model.y0])
inference.run(max_iter=max_iter, learning_rate=learning_rate, verbose=False, callback=print_status, **kwargs)
```

And the inference algorithm would figure out the required subgraph to compute those targets automatically. (Just traversing up the FactorGraph.)

Reply:
Do you mean "targets" here in the sense of classification targets, or inference targets? If it's the former, then I think that's pretty specific to the supervised learning setting. If it's the latter, then actually I think this is quite reasonable in many cases - to be able to specify which variables are observed and which variables you're interested in, and only perform computations on the subgraph that these represent. Note that this is the model that Infer.NET uses (set observed variables, declare which variables you are interested in marginal posteriors for. Of course, sometimes you are truly interested in the full posterior (e.g. when performing MCMC) and this makes less sense.